File: kartwheel/app/src/main/java/us/handstand/kartwheel/model/columnadapter/ColumnAdapters.java
package us.handstand.kartwheel.model.columnadapter;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.gson.reflect.TypeToken;
import com.squareup.sqldelight.ColumnAdapter;

import java.lang.reflect.Type;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import us.handstand.kartwheel.model.Course;
import us.handstand.kartwheel.model.Point;
import us.handstand.kartwheel.network.API;
import us.handstand.kartwheel.util.DateFormatter;

public class ColumnAdapters {
    public static final ColumnAdapter<Course, byte[]> COURSE_BLOB = new ColumnAdapter<Course, byte[]>() {
        @NonNull
        @Override
        public Course decode(byte[] databaseValue) {
            return blobToCourse(databaseValue);
        }

        @Override
        public byte[] encode(@NonNull Course value) {
            return courseToBlob(value);
        }
    };

    @NonNull
    public static byte[] courseToBlob(Course course) {
        if (course == null) {
            return new byte[0];
        } else {
            return API.INSTANCE.getGson().toJson(course).getBytes();
        }
    }

    @NonNull
    public static Course blobToCourse(byte[] blob) {
        if (blob == null || blob.length == 0) {
            return Course.EMPTY_COURSE;
        } else {
            return API.INSTANCE.getGson().fromJson(new String(blob), Course.class);
        }
    }

    public static final ColumnAdapter<Date, Long> DATE_LONG = new ColumnAdapter<Date, Long>() {
        @NonNull
        @Override
        public Date decode(Long databaseValue) {
            Date result = longToDate(databaseValue);
            return result == null ? new Date() : result;
        }

        @Override
        public Long encode(@NonNull Date value) {
            Long result = dateToLong(value);
            return result == null ? 0L : result;
        }
    };

    @Nullable
    public static Date longToDate(Long value) {
        return value == null ? null : DateFormatter.INSTANCE.get(value);
    }

    @Nullable
    public static Long dateToLong(Date date) {
        return date == null ? null : date.getTime();
    }

    public static final ColumnAdapter<List<String>, byte[]> LIST_STRING_BLOB = new ColumnAdapter<List<String>, byte[]>() {
        @NonNull
        @Override
        public List<String> decode(byte[] databaseValue) {
            return blobToListString(databaseValue);
        }

        @Override
        public byte[] encode(@NonNull List<String> value) {
            return listStringToBlob(value);
        }
    };

    private static final Type listStringType = new TypeToken<List<String>>() {
    }.getType();

    @NonNull
    public static List<String> blobToListString(byte[] blob) {
        if (blob == null || blob.length == 0) {
            return Collections.emptyList();
        } else {
            return API.INSTANCE.getGson().fromJson(new String(blob), listStringType);
        }
    }

    @NonNull
    public static byte[] listStringToBlob(List<String> list) {
        if (list == null) {
            return new byte[0];
        } else {
            return API.INSTANCE.getGson().toJson(list).getBytes();
        }
    }

    public static final ColumnAdapter<List<Point>, byte[]> LIST_POINT_BLOB = new ColumnAdapter<List<Point>, byte[]>() {
        @NonNull
        @Override
        public List<Point> decode(byte[] databaseValue) {
            return blobToListPoint(databaseValue);
        }

        @Override
        public byte[] encode(@NonNull List<Point> value) {
            return listPointToBlob(value);
        }
    };

    private static final Type listPointType = new TypeToken<List<Point>>() {
    }.getType();

    @NonNull
    public static List<Point> blobToListPoint(byte[] blob) {
        if (blob == null || blob.length == 0) {
            return Collections.emptyList();
        } else {
            return API.INSTANCE.getGson().fromJson(new String(blob), listPointType);
        }
    }

    @NonNull
    public static byte[] listPointToBlob(List<Point> list) {
        if (list == null) {
            return new byte[0];
        } else {
            return API.INSTANCE.getGson().toJson(list).getBytes();
        }
    }
}

File: kartwheel/app/src/main/kotlin/us/handstand/kartwheel/model/ColumnAdapters.kt
package us.handstand.kartwheel.model


import com.google.gson.reflect.TypeToken
import com.squareup.sqldelight.ColumnAdapter
import us.handstand.kartwheel.network.API
import us.handstand.kartwheel.util.DateFormatter
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.util.*

object ColumnAdapters {
    val DATE_LONG: ColumnAdapter<Date, Long> = object : ColumnAdapter<Date, Long> {
        override fun decode(databaseValue: Long?): Date =
                if (databaseValue != null) DateFormatter[databaseValue] ?: Date() else Date()

        override fun encode(value: Date): Long {
            @Suppress("UNNECESSARY_SAFE_CALL", "USELESS_ELVIS")
            return value?.time ?: 0L
        }
    }

    val LIST_POINT_BLOB: ColumnAdapter<List<Point>, ByteArray> = object : ColumnAdapter<List<Point>, ByteArray> {
        override fun decode(databaseValue: ByteArray): List<Point> = blobToListPoint(databaseValue)

        override fun encode(value: List<Point>): ByteArray = listPointToBlob(value)
    }

    private val listPointType = object : TypeToken<List<Point>>() {

    }.type

    fun blobToListPoint(blob: ByteArray?): List<Point> {
        if (blob == null || blob.isEmpty()) {
            return emptyList()
        } else {
            return API.gson.fromJson<List<Point>>(String(blob), listPointType)
        }
    }

    fun listPointToBlob(list: List<Point>?): ByteArray {
        if (list == null) {
            return ByteArray(0)
        } else {
            return API.gson.toJson(list).toByteArray()
        }
    }

    class ListTicketColumnAdapter<T> : ColumnAdapter<List<T>, ByteArray> {
        override fun decode(databaseValue: ByteArray): List<T> {
            var tickets = emptyList<T>()
            try {
                val `in` = ByteArrayInputStream(databaseValue)
                val `is` = ObjectInputStream(`in`)
                @Suppress("UNCHECKED_CAST")
                tickets = `is`.readObject() as List<T>
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return tickets
        }

        override fun encode(value: List<T>): ByteArray {
            var bytes = ByteArray(0)
            try {
                val out = ByteArrayOutputStream()
                val os = ObjectOutputStream(out)
                os.writeObject(value)
                bytes = out.toByteArray()
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return bytes
        }
    }
}
